src/runtime_src/hip/core/memory.cpp

```diff
@@ -1,11 +1,341 @@
 // SPDX-License-Identifier: Apache-2.0
-// Copyright (C) 2023 Advanced Micro Device, Inc. All rights reserved.
+// Copyright (C) 2024 Advanced Micro Device, Inc. All rights reserved.
 
+#ifdef _WIN32
+#include <Windows.h>
+#include <Memoryapi.h>
+#else
+#include <sys/mman.h>
+#endif
+
+#include "core/common/unistd.h"
 #include "device.h"
 #include "memory.h"
+#include "hip/config.h"
+#include "hip/hip_runtime_api.h"
+
+namespace xrt::core::hip
+{
+
+  memory::memory(size_t sz, std::shared_ptr<xrt::core::hip::device> dev)
+      : m_size(sz), m_type(memory_type::hip_memory_type_device), m_hip_flags(0), m_host_mem(nullptr), m_device(dev), m_bo(nullptr), m_sync_host_mem_required(false)
+  {
+    assert(m_device);
+
+    // TODO: support non-npu device that may require delayed xrt::bo allocation until xrt kernel is created
+    init_xrt_bo();
+  }
+
+  memory::memory(size_t sz, unsigned int flags, std::shared_ptr<xrt::core::hip::device> dev)
+      : m_size(sz), m_type(memory_type::hip_memory_type_host), m_hip_flags(flags), m_host_mem(nullptr), m_device(dev), m_bo(nullptr), m_sync_host_mem_required(false)
+  {
+    assert(m_device);
+
+    switch (m_hip_flags)
+    {
+    case hipHostMallocDefault:
+    case hipHostMallocPortable:
+      // allocate pinned memory on host only, xrt::bo object will not be allocated.
+      m_host_mem = reinterpret_cast<unsigned char *>(aligned_alloc(xrt_core::getpagesize(), m_size));
+      assert(m_host_mem);
+      lock_pages(m_host_mem, m_size);
```

Comment:
we shouldnt be writing our own function to lock pages.
Again, am not 100% sure. lets discuss in next meeting